src/AggregateFunctions/AggregateFunctionMinMaxAny.h

```diff
@@ -488,17 +485,61 @@ struct SingleValueDataString //-V730
 
     void insertResultInto(IColumn & to) const
     {
+        /// Same comment from write() applies here, but in this case we remove the last NULL character as we expect it to come from
+        /// old statuses that always included it
         if (has())
-            assert_cast<ColumnString &>(to).insertData(getData(), size);
-        else
-            assert_cast<ColumnString &>(to).insertDefault();
+        {
+            const char * data = getData();
+            size_t final_size = data[size - 1] != '\0' ? size : size - 1;
+            if (final_size)
+            {
+                assert_cast<ColumnString &>(to).insertData(data, final_size);
+                return;
+            }
+        }
+
+        assert_cast<ColumnString &>(to).insertDefault();
```

Comment:
I would better return old code everywhere and just fix `read` method by appending extra '\0' if and only if data does not end with '\0'. So:
 - Everything will just work if data was written by "old" version.
 - Data that was written by "broken" version will be converted to "old" format on the fly and everything will work*
 - If a string that really ends with '\0' was written by "broken" version, then this '\0' will be trimmed. But it's not a big issue.
 - If an attacker sends data without terminating '\0', it will be sanitized and '\0' will be added (I guess it was the security issue in the old code and that's why the old code was removed)

Reply:
That was my initial approach too. It was a bit more complex since I wanted to deal with the change between internal and external storage efficiently, but maybe it could be simplified.

You can see the code in https://github.com/ClickHouse/ClickHouse/blob/1a8bd114ffc90afb8f2c513aa7f80534627b3dfe/src/AggregateFunctions/AggregateFunctionMinMaxAny.h#L454-L589

Reply:
Yes, it's actually the same approach, but updating only `read` method seems to be simpler and easier to understand. And I don't understand why changes in `write` method are required (any changes in `write` are dangerous, because we may accidentally get third incompatible format)

Reply:
The `write` method was unchanged. I only added comments there.
The `insertResultInto` method did change to avoid passing the ending NULL to the string column (size -1 instead of size) as `insertData` already adds it.

Reply:
Also I think this wasn't the final version I wrote before moving to change the writers instead of the readers. I can try to resurrect that code and clean it up a bit if you prefer.